From e8221b0f3a805de50c94b8c18e5fccef586eb0f8 Mon Sep 17 00:00:00 2001
From: Rob Clark <rob@ti.com>
Date: Mon, 2 Jan 2012 11:57:38 -0600
Subject: [PATCH 154/154] pvrvideosink: fix incorrect window coordinates

When video is larger than screen resolution, the request to create
window at video size will result in a smaller than requested window.
So get the actual window geometry after the window is created.  This
solves an issue with video appearing "split" with the right half of
the video appearing on the left of the window.
---
 sys/pvr2d/gstpvrvideosink.c |   20 ++++++++++----------
 1 files changed, 10 insertions(+), 10 deletions(-)

diff --git a/sys/pvr2d/gstpvrvideosink.c b/sys/pvr2d/gstpvrvideosink.c
index cc841b2..1273320 100644
--- a/sys/pvr2d/gstpvrvideosink.c
+++ b/sys/pvr2d/gstpvrvideosink.c
@@ -633,12 +633,6 @@ gst_pvrvideosink_create_window (GstPVRVideoSink * pvrvideosink, gint width,
   xwindow = g_new0 (GstXWindow, 1);
 
   xwindow->internal = TRUE;
-  pvrvideosink->render_rect.x = pvrvideosink->render_rect.y = 0;
-  pvrvideosink->render_rect.w = width;
-  pvrvideosink->render_rect.h = height;
-  xwindow->width = width;
-  xwindow->height = height;
-  xwindow->internal = TRUE;
 
   g_mutex_lock (dcontext->x_lock);
 
@@ -669,6 +663,8 @@ gst_pvrvideosink_create_window (GstPVRVideoSink * pvrvideosink, gint width,
 
   g_mutex_unlock (dcontext->x_lock);
 
+  gst_pvrvideosink_xwindow_update_geometry (pvrvideosink);
+
   GST_DEBUG_OBJECT (pvrvideosink, "end");
   return xwindow;
 }
@@ -785,7 +781,14 @@ gst_pvrvideosink_blit (GstPVRVideoSink * pvrvideosink, GstBuffer * buffer)
     p_blt_3d->rcSource.bottom = video_height;
   }
 
-  p_blt_3d->hUseCode = NULL;
+  GST_DEBUG_OBJECT (pvrvideosink, "blit: %dx%d (%d) -> %dx%d (%d)",
+      p_blt_3d->sSrc.SurfWidth, p_blt_3d->sSrc.SurfHeight, p_blt_3d->sSrc.Stride,
+      p_blt_3d->sDst.SurfWidth, p_blt_3d->sDst.SurfHeight, p_blt_3d->sDst.Stride);
+  GST_DEBUG_OBJECT (pvrvideosink, "crop: %d,%d %d,%d -> %d,%d %d,%d",
+      p_blt_3d->rcSource.left, p_blt_3d->rcSource.top,
+      p_blt_3d->rcSource.right, p_blt_3d->rcSource.bottom,
+      p_blt_3d->rcDest.left, p_blt_3d->rcDest.top,
+      p_blt_3d->rcDest.right, p_blt_3d->rcDest.bottom);
 
   if (pvrvideosink->format == GST_VIDEO_FORMAT_NV12)
     p_blt_3d->bDisableDestInput = TRUE;
@@ -1126,9 +1129,6 @@ gst_pvrvideosink_setcaps (GstBaseSink * bsink, GstCaps * caps)
         GST_VIDEO_SINK_WIDTH (pvrvideosink),
         GST_VIDEO_SINK_HEIGHT (pvrvideosink));
 
-  pvr_recreate_drawable (pvrvideosink);
-  pvr_get_drawable_params (pvrvideosink);
-
   g_mutex_unlock (pvrvideosink->flow_lock);
 
   gst_pvrvideosink_set_event_handling (GST_X_OVERLAY (pvrvideosink), TRUE);
-- 
1.7.5.4

